garage/logger/singleton_logger.py

```diff
@@ -0,0 +1,99 @@
+"""This is Garage's logger singleton.
+
+It takes in many different types of input and directs them to the correct
+output.
+"""
+from contextlib import contextmanager
+from warnings import warn
+
+
+class Logger():
+    """This is the singleton class that handles logging."""
+
+    def __init__(self):
+        self._outputs = []
+        self._prefixes = []
+        self._prefix_str = ''
+
+    def log(self, data, **kwargs):
+        """Magic method that takes in all different types of input."""
+        if not self._outputs:
+            warn('No outputs have been added to the logger.')
+
+        kwargs['prefix'] = self._prefix_str
+        if 'with_prefix' in kwargs and not kwargs['with_prefix']:
+            kwargs['prefix'] = ''
+
+        for output in self._outputs:
+            if isinstance(data, output.types_accepted):
+                output.log_output(data, **kwargs)
+
+    def add_output(self, output):
+        """Add a new output to the logger.
+
+        All data that is compatible with this output will be sent there.
+        """
+        self._outputs.append(output)
+
+    def remove_all(self):
+        """Remove all outputs that have been added to this logger."""
+        self._outputs.clear()
+
+    def remove_output(self, output_type):
+        """Remove all outputs of a given type."""
+        self._outputs = [
+            output for output in self._outputs
+            if not isinstance(output, output_type)
+        ]
+
+    def reset_output(self, output):
+        """Removes, then re-adds a given output to the logger."""
+        self.remove_output(type(output))
+        self.add_output(output)
+
+    def has_output(self, output_type):
+        """Checks to see if a given logger output is attached to the logger."""
+        for output in self._outputs:
+            if isinstance(output, output_type):
+                return True
+        return False
+
+    def get_output(self, output_type):
```

Comment:
Please don't clobber me previous comments about how you address the log outputs here. It seems strange to address them on type which I could completely validly want 2 CSV outputs.

Reply:
@krzentner @naeioi

Reply:
I have removed this method in an upcoming commit